lib/websocket/websocket.js

```diff
@@ -218,25 +235,72 @@ class WebSocket extends EventTarget {
+      this.#bufferedAmount += buffer.byteLength
+      socket.write(buffer)
+    } else if (types.isArrayBuffer(data)) {
+      // If the WebSocket connection is established, and the WebSocket
+      // closing handshake has not yet started, then the user agent must
+      // send a WebSocket Message comprised of data using a binary frame
+      // opcode; if the data cannot be sent, e.g. because it would need
+      // to be buffered but the buffer is full, the user agent must flag
+      // the WebSocket as full and then close the WebSocket connection.
+      // The data to be sent is the data stored in the buffer described
+      // by the ArrayBuffer object. Any invocation of this method with an
+      // ArrayBuffer argument that does not throw an exception must
+      // increase the bufferedAmount attribute by the length of the
+      // ArrayBuffer in bytes.
+
+      frame.opcode = opcodes.BINARY
+      frame.payloadData = Buffer.from(data)
+
+      const buffer = frame.toBuffer()
+      this.#bufferedAmount += buffer.byteLength
+      socket.write(buffer)
+    } else if (ArrayBuffer.isView(data)) {
+      // If the WebSocket connection is established, and the WebSocket
+      // closing handshake has not yet started, then the user agent must
+      // send a WebSocket Message comprised of data using a binary frame
+      // opcode; if the data cannot be sent, e.g. because it would need to
+      // be buffered but the buffer is full, the user agent must flag the
+      // WebSocket as full and then close the WebSocket connection. The
+      // data to be sent is the data stored in the section of the buffer
+      // described by the ArrayBuffer object that data references. Any
+      // invocation of this method with this kind of argument that does
+      // not throw an exception must increase the bufferedAmount attribute
+      // by the length of data’s buffer in bytes.
+
+      frame.opcode = opcodes.BINARY
+      frame.payloadData = Buffer.from(data)
+
+      const buffer = frame.toBuffer()
+      this.#bufferedAmount += buffer.byteLength
+      socket.write(buffer)
+    } else if (isBlobLike(data)) {
+      // If the WebSocket connection is established, and the WebSocket
+      // closing handshake has not yet started, then the user agent must
+      // send a WebSocket Message comprised of data using a binary frame
+      // opcode; if the data cannot be sent, e.g. because it would need to
+      // be buffered but the buffer is full, the user agent must flag the
+      // WebSocket as full and then close the WebSocket connection. The data
+      // to be sent is the raw data represented by the Blob object. Any
+      // invocation of this method with a Blob argument that does not throw
+      // an exception must increase the bufferedAmount attribute by the size
+      // of the Blob object’s raw data, in bytes.
+
+      frame.opcode = opcodes.BINARY
+
+      data.arrayBuffer().then((ab) => {
```

Comment:
If you have concurrent writes this might change the order of messages:

```js
websocket.send(blob);
websocket.send(string);
```

The string message is sent before the blob.

Reply:
I don't know of any good way of fixing this without needing a Promise for every send.. Preferably node would allow writing a blob to a writable stream.

Reply:
This is the reason why there is no `Blob` support in `ws`. This and because the user can already do

```js
websocket.send(await blob.arrayBuffer());
```

I agree with https://github.com/ricea/websocketstream-explainer/blob/832f88d7f8eaecbe51da318d802951a3f8aadb2f/README.md?plain=1#L152-L158

Reply:
Yes, when I was writing out my comment is kinda just clicked - "oh... this is why ws doesn't support blobs", lol. I still think undici should implement it for the sake of complying with the spec.

When I get around to implementing websocketstream this shouldn't be a problem at least!

Reply:
Well, `ws` already uses a queue (only when `permessage-deflate` is enabled and negotiated since compression is asynchronous) so support for `Blob` could be added, but I think it does not worth the effort. I think that `Blob` is of little use in Node.js if not natively supported by streams.

Reply:
hmm, but `send(await blob.arrayBuffer())` would require reading the hole blob into memory before sending...
sure you could do something like: 
```js
for await (const chunk of blob.stream()) {
  websocket.send(chunk)
}
```
But then you would get multiple messages and you would have to assemble them yourself on the receiving side as well 😞 

if you can accept blobs then it could as well be piped to a tmp location on the disc and wouldn't hold up any memory when receiving blobs

Reply:
Ah, just went into the github repo and learned about a new `WebSocketStream` forgot that it existed, only head about it when it was in early proposal. Well, then you could instead maybe do: `blob.stream().pipeTo(websocket)`?

Reply:
> But then you would get multiple messages and you would have to assemble them yourself on the receiving side as well 😞

That's up you. With `ws` you could send each chunk as a message fragment and receive the whole thing as a single message on the receiving end.

```js
for await (const chunk of blob.stream()) {
  websocket.send(chunk, { fin: false}); // For the sake of simplicity backpressure is not handled.
}
websocket.send(Buffer.alloc(0), { fin: true });
```

> if you can accept blobs then it could as well be piped to a tmp location on the disc and wouldn't hold up any memory when receiving blobs

AFAIK the whole message is buffered in memory before the `'message'` event is emitted. That message is then returned to the user as a `Blob`.